Source/WebCore/rendering/RenderBox.cpp

```diff
@@ -1394,6 +1394,85 @@ void RenderBox::clearOverridingLogicalWidthLength()
         gOverridingLogicalWidthLengthMap->remove(this);
 }
 
+FlowRelativeDirection RenderBox::physicalToFlowRelativeDirectionMapping(PhysicalDirection direction) const
+{
+    auto isHorizontalWritingMode = style().isHorizontalWritingMode();
+    auto isLeftToRightDirection = style().isLeftToRightDirection();
+
+    if (isHorizontalWritingMode == containingBlock()->style().isHorizontalWritingMode()) {
+        switch (direction) {
+        case PhysicalDirection::Top:
+            return isHorizontalWritingMode ? FlowRelativeDirection::BlockStart : (isLeftToRightDirection ? FlowRelativeDirection::InlineStart : FlowRelativeDirection::InlineEnd);
+        default:
+            ASSERT_NOT_IMPLEMENTED_YET();
+        } 
+    } else
+        ASSERT_NOT_IMPLEMENTED_YET();
+    return { };
+}
+
+static MarginTrimType flowRelativeDirectionToMarginTrimType(FlowRelativeDirection direction)
+{
+    switch (direction) {
+    case FlowRelativeDirection::BlockStart:
+        return MarginTrimType::BlockStart;
+    case FlowRelativeDirection::BlockEnd:
+        return MarginTrimType::BlockEnd;
+    case FlowRelativeDirection::InlineStart:
+        return MarginTrimType::InlineStart;
+    case FlowRelativeDirection::InlineEnd:
+        return MarginTrimType::InlineEnd;
+    default:
+        ASSERT_NOT_REACHED();
+        return { };
+    }
+}
+
+void RenderBox::markMarginAsTrimmed(MarginTrimType newTrimmedMargin)
+{
+    ensureRareData().setTrimmedMargins(rareData().trimmedMargins() | static_cast<unsigned>(newTrimmedMargin));
+}
+
+void RenderBox::clearTrimmedMarginsMarkings()
+{
+    ASSERT(hasRareData());
+    ensureRareData().setTrimmedMargins(0);
+    setNeedsLayout(MarkOnlyThis);
```

Comment:
Could you explain the logic behind this call, please? To my understanding clearTrimmedMarginsMarkings() is about invalidating a cached value of whether the associated box has trimmed margins, right? This value is valid from the point of computing the trimmed value (if applicable) until we believe the renderer needs new geometry (for whatever reason). It makes me believe that at the point of the invalidation the associated box should already be marked damaged and if not, the callsite should be doing that instead (it has more context to decide the extent of the damage as calling setNeedsLayout only on *this does not change the flow of the layout e.g. the renderer may very well stay damaged)

Reply:
My idea is that these 2 things (unmarking the trimming bits on a child and laying it out again) would always go together and am not sure why clearing the bits would not be followed with a layout. However, I do see you point about the caller having more context about this and my idea was that it would only be used in those scenarios where the caller was attempting to perform a layout (probably because it is involved with some sort of formatting context), of some children.

I could also see how someone could accidentally call this method in a scenario where they probably don't mean to and find themselves with a box that was marked for layout. 

I went ahead and removed the call from the function and moved it to the call site of `clearTrimmedMarginsMarkings ` instead.

Reply:
We normally mark renderers dirty when
1, some kind of style or content mutation happens (before layout)
2, need to expand the damage on descendant renders (or siblings/nephews etc in rare cases) during layout because constraints etc change and we believe it may have impact on other renderers.
(The second needs carful handling because layout has a well defined flow and as I said earlier marking renderers dirty has no impact on the direction of the flow)
Trimmed margin on a flex item itself should not trigger layout. If margin changes on the child, yes we need to invalidated this value (child is already marked dirty), if flex container's margin-trim value changes, yes we need to invalidate and mark flex items dirty) etc. 
I think the dependency is the other way around, i.e. when the child needs layout, the cached value needs to be invalidated as well.
(does this need to be addressed in this PR?)

Reply:
Alternatively you can introduce an invalid state and compute the cached value lazily on the first call.